include/boundary_op.hxx

```diff
@@ -73,13 +66,49 @@ public:
 
   BoundaryRegion *bndry;
   bool apply_to_ddt; // True if this boundary condition should be applied on the time derivatives, false if it should be applied to the field values
+
+protected:
+  std::shared_ptr<FieldGenerator> gen; // Generator
+
+  // Apply boundary condition at a point
+  virtual void applyAtPoint(Field2D &UNUSED(f), BoutReal UNUSED(val), int
+      UNUSED(x), int UNUSED(bx), int UNUSED(y), int UNUSED(by), int UNUSED(z),
+      Coordinates* UNUSED(metric)) {
+    ASSERT1(false);
```

Comment:
This might lead to a confused user as the error message won't say what is wrong just that the assert failed so a throw might be better. 

I think if you want it to be an error to not override these routines then you can make these be pure virtual, e.g. 

~~~~
virtual void applyAtPoint(Field2D &f, BoutReal val, int x, int bx, int y, int by, int z,
      Coordinates* metric) = 0;
~~~~

and it's then a compile time error.

I notice you pass in the metric here, but you're also passing a field -- could we just get the metric from the field?

Reply:
I implemented like this because subclasses don't have to implement these methods if they override `apply()`. I didn't want to have to put some sort of null version of these four methods in subclasses that don't need them, so had to provide a default implementation. Having at least a helpful error message is a good call.

Another possibilty would be to take `applyAtPoint`, `applyAtPointStaggered` and `extrapPoint` and put them in a subclass like `BoundaryOpWithApply` that inherits from `BoundaryOp` and implements the generic `apply` method. Then the actual implementations could inherit from `BoundaryOp` if they provide `apply` or from `BoundaryOpWithApply` and provide the low-level methods.

I think the `BoundaryOpWithApply` subclass could be a case for using the CRTP design https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern; CRTP is extra code and (to me at least) adds work for the reader to understand, but might have a performance benefit: [from the wikipedia article] _"However, if base class member functions use CRTP for all member function calls, the overridden functions in the derived class will be selected at compile time. This effectively emulates the virtual function call system at compile time without the costs in size or function call overhead (VTBL structures, and method lookups, multiple-inheritance VTBL machinery) at the disadvantage of not being able to make this choice at runtime."_

Any preferences for:
1. Leaving all the methods in `BoundaryOp`
2. Providing a default `apply` method in an intermediate subclass `BoundaryOpWithApply` so that implementations either subclass `BoundaryOp` and implement `apply` or subclass `BoundaryOpWithApply` and provide `applyAtPoint`, etc.
3. Use CRTP to (if I understand right) avoid cost of virtual methods in 2.
4. Something else?

Reply:
> I notice you pass in the metric here, but you're also passing a field -- could we just get the metric from the field?

Good point. Not sure if there are performance implications of having the f.getCoordinates() call on the inside of the loop? Passing `Coordinates* metric` as an argument will be gone in an update anyway, because it turned out to be convenient to calculate a grid spacing outside and pass that instead.